[MCA] Extend -instruction-tables option with verbosity levels

jvillette38 opened the pull request:
Option becomes: -instruction-tables=`<level>`
 
The choice of `<level>` controls number of printed information.
`<level>` may be `none` (default), `normal`, `full`.
Note: If the option is used without `<label>`, default is `normal` (legacy).

When `<level>` is `full`, additional information are:
  - `<Bypass Latency>`: Latency when a bypass is implemented between operands
  in pipelines (see SchedReadAdvance).
  - `<LLVM Opcode Name>`: mnemonic plus operands identifier.
  - `<Resources units>`: Used resources associated with LLVM Opcode.
  - `<instruction comment>`: reports comment if any from source assembly.

Level `full` can be used to better check scheduling info when TableGen is modified.
LLVM Opcode name help to find right instruction regexp to fix TableGen Scheduling Info.

-instruction-tables=full option is validated on AArch64/Neoverse/V1-sve-instructions.s

Follow up of MR #126703

JulienVillette committed:
[MCA] New option -scheduling-info

Outputs micro ops, latency, bypass latency, throughput, llvm opcode
name, used resources and parsed assembly instruction with comments.

This option is used better check scheduling info when TableGen is
modified. Information like Throughput (not reverse throughput)
and bypass latency help to compare with micro architecture
documentation.

LLVM Opcode name help to find right instruction regexp to fix
TableGen Scheduling Info.

Example:
  Input:      abs D20, D11  // ABS <V><d>, <V><n>
      \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV
  Output:  1 2 2 4.00 V1UnitV ABSv1i64 abs     d20, d11
      // ABS <V><d>, <V><n>  \\ ASIMD arith, basic
      \\ 1 2  2  4.0 V1UnitV

RKSimon commented:
Isn't this what -instruction-tables is for?

JulienVillette committed:
[MCA] Fix formatting

JulienVillette committed:
[MCA] Fixing Sphinx formating

JulienVillette committed:
[MCA] Simplify instruction field

Output only Instructions: like for -instruction-tables option

JulienVillette committed:
[MCA] Fix formatting

adibiagio commented:
> > Isn't this what -instruction-tables is for?
> 
> But does not gives same information... You'd like to change output of -instruction-tables? -scheduling-info should be used by llvm developer and -instruction-tables by application developer.

I think Simon has a point.


I think that information like "resource names" and "bypass latency" is also relevant to application developers.
The instruction-tables view was originally inspired by Agner Fog's instruction_tables document. Things like resource names and bypass latency are a good fit for that view.

It would be nicer if your new logic was built on top of the existing instruction tables. At the moment it is designed as a replacement for the instruction info (which personally I don't particularly like).

You could also have an option for printing "throughput" values instead of "reciprocal throughput". 
Anything else, could be just new fields of instruction-tables.

> 
> -scheduling-info output: `Scheduling Info:` `[1]: #uOps` `[2]: Latency` `[3]: Bypass Latency` `[4]: Throughput` `[5]: Resources` `[6]: LLVM OpcodeName` `[1] [2] [3] [4] [5] [6] Instructions:` ` 1 5 2 2.00 V1UnitV,V1UnitV01 BFMLALB_ZZZ bfmlalb z0.s, z1.h, z2.h`
> 
> -instruction-tables output: `Instruction Info:` `[1]: #uOps` `[2]: Latency` `[3]: RThroughput` `[4]: MayLoad` `[5]: MayStore` `[6]: HasSideEffects (U)` `[1] [2] [3] [4] [5] [6] Instructions:` ` 1 5 0.50 bfmlalb z0.s, z1.h, z2.h`

JulienVillette committed:
[MCA] Fix test after simplified instruction field

boomanaiden154 reviewed (commented):
I agree that if we want something in `llvm-mca` that is user facing, this should go under `-instruction-tables`.

If this is intended to be purely developer facing for validating the scheduling models, I still think that could be better done within/around `llvm-exegesis`.

jvillette38 commented:
> Agner Fog's instruction_tables document

Thank you all for the comments.
I didn't know this document.
In fact, this patch match with tables entries seen in it:

| Instruction | Operands | Ops | Latency | Reciprocal throughput | Execution unit | Notes |
|-------------|------------|------|----------|--------------------------|-----------------|--------|
| MOV | r,r | 1 | 1 | 1/3 | ALU |
| MOV | r,i | 1 | 1 | 1/3 | ALU |

Some of additional information are in this patch:
- `Instruction + operands`: LLVM Opcode Name.
- Execution unit

Adding bypass latency.

I am going to write these informations only when -instruction-tables is on.

JulienVillette committed:
[MCA] Print scheduling-info + instruction-tables

adibiagio commented:
> I agree that if we want something in `llvm-mca` that is user facing, this should go under `-instruction-tables`.
> 
> If this is intended to be purely developer facing for validating the scheduling models, I still think that could be better done within/around `llvm-exegesis`.

I agree with you. llvm-exegesis is a better fit for validating and building scheduling models.

Speaking about instruction-tables only: I think it doesn't hurt if we add a verbose option for printing extra fields. Some of that information is still very useful to have in general.

adibiagio commented on llvm/docs/CommandGuide/llvm-mca.rst:
```diff
@@ -170,6 +170,20 @@ option specifies "``-``", then the output will also be sent to standard output.
   Enable extra scheduler statistics. This view collects and analyzes instruction
   issue events. This view is disabled by default.
 
+.. option:: -scheduling-info
+
+  Enable scheduling info view. This view reports scheduling information defined
+  in LLVM target description in the form:
+  ``<uOps> <Latency> <Bypass Latency> <Throughput> <LLVM Opcode Name> <Resources
+  units> <instruction and comment (// or /* */) if defined>``
+  It allows to compare scheduling info with architecture documents and fix them
+  in target description by fixing InstrRW for the reported LLVM opcode.
+  Scheduling information can be defined in the same order in each instruction
+  comments to check easily reported and reference scheduling information.
+  Suggested information in comment:
+  ``// <architecture instruction form> \\ <scheduling documentation title> \\
+  <uOps>, <Latency>, <Bypass Latency>, <Throughput>, <Resources units>``
+
```
I wonder whether we could reuse flag `-instruction-tables`.
e.g.:
```
 -instruction-tables    // print the default view (no extra info). This is the "normal" behavior.
 -instruction-tables=normal    // same as above
 -instruction-tables=full  // print the full view (this includes your extra fields)
```

jvillette38 replied:
> I wonder whether we could reuse flag `-instruction-tables` instead of adding a new flag. e.g.:
> 
> ```
>  -instruction-tables    // print the default view (no extra info). This is the "normal" behavior.
>  -instruction-tables=normal    // same as above
>  -instruction-tables=full  // print the full view (this includes your extra fields)
> ```

Done in last commits. `-instruction-tables` is equivalent to `-instruction-tables=normal` to avoid to modify all tests.

adibiagio commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -757,7 +764,7 @@ int main(int argc, char **argv) {
       Printer.addView(std::make_unique<mca::InstructionInfoView>(
           *STI, *MCII, CE, ShowEncoding, Insts, *IP, LoweredSequence,
-          ShowBarriers, *IM, InstToInstruments));
+          ShowBarriers, ShowSchedulingInfo, *IM, InstToInstruments));
```
You should only print extra info if we are emitting instruction tables.

adibiagio commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -14,10 +14,121 @@
+    FOS.PadToColumn(paddings[1]);
+    FOS << " " << IIVDEntry.Bypass;
+    FOS.PadToColumn(paddings[2]);
+    if (IIVDEntry.RThroughput) {
+      double RT = 1.0 / *IIVDEntry.RThroughput;
+      FOS << " " << format("%.2f", RT);
+    } else {
+      FOS << " -";
+    }
```
Maybe add a flag for controlling whether to print it as reciprocal throughput (default) or normal throughput.

jvillette38 replied:
I will let reverse thourghput ;) Thanks

adibiagio reviewed (commented):
I left a few comment.

mshockwave reviewed (commented):
I also believe that we should piggy-back the new features proposed here on the existing `-instruction-tables` flag

JulienVillette committed:
[MCA] Rework InstructionInfoView printing to be able to add full view

Option instruction-tables is an enumeration now but
legacy behavior (boolean) is kept available.
New option: -instruction-tables=<none,normal,full>.
-instruction-tables is equivalent to -instruction-tables=normal.
-instruction-tables=none is the default.
-instruction-tables=full is equivalent to:
 -instruction-tables and -scheduling-info
-sheduling-info option removed.

JulienVillette committed:
[MCA] Replace scheduling info pring by print full instruction tables view

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -14,13 +14,45 @@
+void InstructionInfoView::getComment(const MCInst &MCI,
+                                     std::string &CommentString) const {
+  StringRef s = MCI.getLoc().getPointer();
+  std::string InstrStr;
```
judging from your later uses, I think we can use `StringRef` here. `std::string` is usually (much) more expensive than `StringRef`

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -14,13 +14,45 @@
+    }
+    // C++ style comment
+    if ((pos_cmt = InstrStr.find("//")) != std::string::npos) {
+      CommentString = InstrStr.substr(pos_cmt, pos);
```
I'm not sure why you put `pos` here: if `pos` was not updated by line 38, then it'll be a position of _another_ string, `s`, which is not ideal. Since you already chop off new line character from `s`, I don't think `InstrStr` has any of them. So I think you can just use `InstrStr.substr(pos_cmt)` here.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+        FOS.PadToColumn(20);
+        for (unsigned U = 0; U < NumUnits; ++U) {
+          FOS << SM.getProcResource(ProcResource.SubUnitsIdxBegin[U])->Name;
+          if ((U + 1) < NumUnits) {
```
drop curly braces

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+    }
+  }
+
+  std::vector<unsigned> paddings = {0, 7, 14, 21, 28, 35};
+  std::vector<std::string> fields = {"#uOps",       "Latency",
```
Please capitalize the variable names.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+  std::vector<std::string> fields = {"#uOps",       "Latency",
+                                     "RThroughput", "MayLoad",
+                                     "MayStore",    "HasSideEffects (U)"};
+  std::vector<std::string> end_fields;
```
ditto variable names

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+  }
+
+  FOS << "\n\nInstruction Info:\n";
+  for (unsigned i = 0; i < fields.size(); i++) {
```
```suggestion
  for (unsigned i = 0, N = fields.size(); i < N; i++) {
```

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+  }
+  FOS << "\n";
+
+  for (unsigned i = 0; i < paddings.size(); i++) {
```
ditto `padding.size()`

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+  FOS << "\n";
+
+  for (unsigned i = 0; i < paddings.size(); i++) {
+    if (paddings[i] != 0)
```
```suggestion
    if (paddings[i])
```

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+      FOS.PadToColumn(paddings[i]);
+    if (i < fields.size()) {
+      FOS << "[" << i + 1 << "]";
     } else {
-      TempStream << "\n[1]    [2]    [3]    [4]    [5]    [6]    "
-                 << "Instructions:\n";
+      FOS << end_fields[i - fields.size()];
     }
```
drop curly braces

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,152 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+    FOS.PadToColumn(paddings[LastPaddingIdx + 1]);
+    FOS << printInstructionString(Inst);
+    if (PrintFullInfo) {
+      getComment(Inst, CommentString);
```
nit: can we move the declaration of `CommentString` to here, I think these two lines are the only users.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.h:
```diff
@@ -65,28 +66,39 @@ class InstructionInfoView : public InstructionView {
     bool mayLoad = false;
     bool mayStore = false;
     bool hasUnmodeledSideEffects = false;
+    std::string OpcodeName = "";
```
judging from your uses, this is storing opcode names returned from `MCInstrInfo::getName`: that function gives you a string from a giant `static` table, which is why it returns a `StringRef`. So I think you can use `StringRef` to avoid extra allocation induced by `std::string`.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +243,34 @@ void InstructionInfoView::collectData(
+      // Get latency with bypass
+      IIVDEntry.Bypass =
+          IIVDEntry.Latency - MCSchedModel::getBypassDelayCycles(STI, SCDesc);
+      IIVDEntry.OpcodeName = (std::string)MCII.getName(Inst.getOpcode());
```
please avoid casting to `std::string`, after you change `OpcodeName` to StringRef.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +243,34 @@ void InstructionInfoView::collectData(
+      raw_string_ostream TempStream(IIVDEntry.Resources);
+      const MCWriteProcResEntry *Index = STI.getWriteProcResBegin(&SCDesc);
+      const MCWriteProcResEntry *Last = STI.getWriteProcResEnd(&SCDesc);
+      auto sep = "";
```
Please capitalize the variable name and spell out the type

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +243,34 @@ void InstructionInfoView::collectData(
+        }
+        sep = ",";
+      }
+      TempStream.flush();
```
`raw_string_ostream`, unlike `std::stringstream`, is recommended not to flush it manually.

mshockwave commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -364,6 +405,8 @@ int main(int argc, char **argv) {
 
   cl::HideUnrelatedOptions({&ToolOptions, &ViewOptions});
 
+  InstructionTablesOption.setValueExpectedFlag(cl::ValueOptional);
```
please just add `cl::ValueOptional` in the declaration of `InstructionTablesOption`

mshockwave commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,51 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
-                           cl::cat(ToolOptions), cl::init(false));
+enum class InstructionTablesType { NONE, NORMAL, FULL };
+
+class InstructionTablesOptionParser
+    : public cl::parser<enum InstructionTablesType> {
+public:
+  explicit InstructionTablesOptionParser(cl::Option &O)
+      : cl::parser<enum InstructionTablesType>(O) {}
+
+  bool parse(cl::Option &O, StringRef ArgName, StringRef Arg,
+             enum InstructionTablesType &Value) {
+    if (Arg.empty()) {
+      Value = InstructionTablesType::NORMAL;
+      return false;
+    }
+    return cl::parser<enum InstructionTablesType>::parse(O, ArgName, Arg,
+                                                         Value);
+  }
+};
```
instead of using custom parser, I would recommend adding a `cl::EnumValN(InstructionTablesType::NORMAL, "", "")` to the list of `cl::values` in the declaration of `InstructionTablesOption`. So that it'll choose `NORMAL` by default in the absent of value.

mshockwave commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,51 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
+                              " used resources")),
+        cl::cat(ToolOptions), cl::init(InstructionTablesType::NONE));
+
+bool PrintInstructionTables() {
+  if (InstructionTablesOption == InstructionTablesType::NONE)
+    return false;
+
+  return true;
+}
```
```suggestion
bool PrintInstructionTables() {
  return PrintInstructionTables(InstructionTablesType::NONE);
}
```

mshockwave commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,51 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
+  return true;
+}
+
+bool PrintInstructionTables(enum InstructionTablesType ITType) {
```
please mark these two `PrintInstructionTables` as static. Also, since you'd turned it into a function, it has to follow the function naming convention, which starts with a lower case.

mshockwave commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -757,7 +802,7 @@ int main(int argc, char **argv) {
       Printer.addView(std::make_unique<mca::InstructionInfoView>(
           *STI, *MCII, CE, ShowEncoding, Insts, *IP, LoweredSequence,
-          ShowBarriers, *IM, InstToInstruments));
+          ShowBarriers, false, *IM, InstToInstruments));
```
```suggestion
          ShowBarriers, /*ShouldPrintFullInfo=*/false, *IM, InstToInstruments));
```

michaelmaitland commented:
Does the title need to reflect that we're now using `-instruction-tables`?

JulienVillette committed:
[MCA] Update after review

jvillette38 commented:
> Does the title need to reflect that we're now using `-instruction-tables`?

Thank you for the review. Pushed changes.

JulienVillette committed:
[MCA] Update after review

michaelmaitland commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,32 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
+                          " used resources")),
+    cl::cat(ToolOptions), cl::init(InstructionTablesType::NONE),
+    cl::ValueOptional);
+
```
nit: the naming of this sounds like it prints the instruction tables. rename to `shouldPrintInstructionTables`?

michaelmaitland commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +241,33 @@ void InstructionInfoView::collectData(
+          continue;
+        const MCProcResourceDesc *MCProc =
+            SM.getProcResource(Index->ProcResourceIdx);
+        if (Index->ReleaseAtCycle > 1) {
```
Need to handle AcquireAtCycles?

michaelmaitland replied:
Thanks for making the update here.

 > https://github.com/llvm/llvm-project/blob/main/llvm/lib/MC/MCSchedule.cpp#L97 shows that AcquireAtCycle is not considered when computing throughput.

That might be a bug? @mshockwave what do you think?

mshockwave replied:
> Thanks for making the update here.
> 
> > https://github.com/llvm/llvm-project/blob/main/llvm/lib/MC/MCSchedule.cpp#L97 shows that AcquireAtCycle is not considered when computing throughput.
> 
> That might be a bug? @mshockwave what do you think?

Yeah I tried to fix it but that (accounting for AcquireAtCycle or the concept of ResourceSegment in general) turns out to be more complicated than I thought: we can't just use the smallest `NumUnit / (ReleaseAtCycle - AcquireAtCycle)` to calculate the (inverse) throughput

jvillette38 replied:
In practice, do you have `AcquireAtCycle` not defined to 0 for at least one resource?
I think it has no sense to define `AcquireAtCycle` != 0 on all resources. So, throughput depends only on `ReleaseAtCycle`.
Example `vlm.v`: `SiFive7VCQ,SiFive7VL[1,3]`. First resource has `AcquireAtCycle` == 0 and `ReleaseAtCycle` == 1.
Are you agree with that?

jvillette38 replied:
Your are right! So throughput is not constant...
```
          |  0  |  1  |  2  |  3  |  4  |  5  |  6  |  7  |  8  |  9  |
SiFive7CQ    X           Y           Z           W
SiFive7VL          X     X     Y     Y     Z     Z     W     W
```
- 3 cycles for 1 instruction: RThroughput = 3
- 5 cycles for 2 instructions: RThroughput = 2.5
- 7 cycles for 3 instructions: RThroughput = 2.33
- 9 cycles for 4 instructions: RThroughput =  2.25
Instructions = x
Cycles = (x * 2) + 1
Cycles = (x * (ReleaseAtCycle - AcquireAtCycle)) + AcquireAtCycle
RThroughput =  [(x * (ReleaseAtCycle - AcquireAtCycle)) + AcquireAtCycle] / x

Are you agree with this formula?
But it does not solve our issue to define constant throughput...

![image](https://github.com/user-attachments/assets/c6759d35-579d-4666-81d6-95d4ce5da968)

Do we consider large number of instructions and the limit +oo? So in this case, it would give a throughput of 2 = (ReleaseAtCycle - AcquireAtCycle).
And if we have more than 1 resource: Reverse Throughput will be (ReleaseAtCycle - AcquireAtCycle) / Number of resources.

Finally: is it better to consider throughput of one instruction or lot of identical instructions?
Note: the max error is equal to AcquireAtCycle.

jvillette38 replied:
We should modify `MCSchedModel::getReciprocalThroughput(const MCSubtargetInfo`:
- Version that returns RThroughput for big number of identical instructions:
```CPP
double
MCSchedModel::getReciprocalThroughput(const MCSubtargetInfo &STI,
                                      const MCSchedClassDesc &SCDesc) {
  std::optional<double> Throughput;
  const MCSchedModel &SM = STI.getSchedModel();
  const MCWriteProcResEntry *I = STI.getWriteProcResBegin(&SCDesc);
  const MCWriteProcResEntry *E = STI.getWriteProcResEnd(&SCDesc);
  for (; I != E; ++I) {
    if (!I->ReleaseAtCycle)
      continue;
    unsigned NumUnits = SM.getProcResource(I->ProcResourceIdx)->NumUnits;
    unsigned Range =  I->ReleaseAtCycle;
    if (I->AcquireAtCycle)
      Range = Range - AcquireAtCycle;
    double Temp = NumUnits * 1.0 / Range;
    Throughput = Throughput ? std::min(*Throughput, Temp) : Temp;
  }
  if (Throughput)
    return 1.0 / *Throughput;

  // If no throughput value was calculated, assume that we can execute at the
  // maximum issue width scaled by number of micro-ops for the schedule class.
  return ((double)SCDesc.NumMicroOps) / SM.IssueWidth;
}
```
- Version that returns RThroughput for one instruction but considering max range [AcquireAtCycle,ReleaseAtCycle]:
```CPP
double
MCSchedModel::getReciprocalThroughput(const MCSubtargetInfo &STI,
                                      const MCSchedClassDesc &SCDesc) {
  std::optional<double> Throughput;
  const MCSchedModel &SM = STI.getSchedModel();
  const MCWriteProcResEntry *I = STI.getWriteProcResBegin(&SCDesc);
  const MCWriteProcResEntry *E = STI.getWriteProcResEnd(&SCDesc);
  unsigned MaxRange = 0;
  for (; I != E; ++I) {
    if (!I->ReleaseAtCycle)
      continue;
    unsigned NumUnits = SM.getProcResource(I->ProcResourceIdx)->NumUnits;
    unsigned Range =  I->ReleaseAtCycle;
    if (I->AcquireAtCycle)
      Range = Range - AcquireAtCycle;
    if (Range > MaxRange) {
      MaxRange = Range;
      double Temp = NumUnits * 1.0 / I->ReleaseAtCycle;
      Throughput = Throughput ? std::min(*Throughput, Temp) : Temp;
    }
  }
  if (Throughput)
    return 1.0 / *Throughput;

  // If no throughput value was calculated, assume that we can execute at the
  // maximum issue width scaled by number of micro-ops for the schedule class.
  return ((double)SCDesc.NumMicroOps) / SM.IssueWidth;
}
```

mshockwave replied:
Alright, you nerd-sniped me and I've proved that we should indeed use the largest `(ReleaseAtCycle - AcquireAtCycle) / NumUnits` as inverse throughput...but the reason behind it is not so simple: https://myhsu.xyz/llvm-sched-interval-throughput
I've also created #132653 to update the formula (with you as co-author, ofc)

jvillette38 replied:
Thanks! I am agree with you. The goal is to get the minimum throughput and so maximum reverse throughtput of the instruction to be used for example in cost model.
I was thinking about that these last days and so I would propose the same patch than your MR #132653.

JulienVillette committed:
[MCA] Update after review

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +60,151 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+  unsigned LastPadding = 35;
+  if (PrintFullInfo) {
+    Fields.push_back("Bypass Latency");
+    Paddings.push_back(LastPadding + 7);
+    LastPadding += 7;
```
@topperc spot this yesterday:
```suggestion
    LastPadding += 7;
    Paddings.push_back(LastPadding);
```
ditto every other similar cases in this file

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +241,33 @@ void InstructionInfoView::collectData(
+          continue;
+        const MCProcResourceDesc *MCProc =
+            SM.getProcResource(Index->ProcResourceIdx);
+        if (Index->ReleaseAtCycle > 1) {
```
Did you address @michaelmaitland 's comment on AcquireAtCycles?

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +241,33 @@ void InstructionInfoView::collectData(
+          TempStream << Sep
+                     << format("%s[%d]", MCProc->Name, Index->ReleaseAtCycle);
+        } else {
+          TempStream << Sep << format("%s", MCProc->Name);
```
I think a plain `TempStream << Sep << MCProc->Name` should be sufficient here

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -141,6 +241,33 @@ void InstructionInfoView::collectData(
+      raw_string_ostream TempStream(IIVDEntry.Resources);
+      const MCWriteProcResEntry *Index = STI.getWriteProcResBegin(&SCDesc);
+      const MCWriteProcResEntry *Last = STI.getWriteProcResEnd(&SCDesc);
+      auto Sep = "";
```
I don't think you addressed my previous comment on this line.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +60,151 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
   }
 
-  TempStream.flush();
+  FOS.flush();
```
I don't think you addressed my previous comment on this line.

mshockwave commented:
Please avoid force pushing as much as possible. Some comments might be lost after forced push.

michaelmaitland commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,32 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
+  return false;
+}
+
+static bool printInstructionTables() {
```
rename here too for same reasons as above? This function does not actually print any tables

JulienVillette committed:
[MCA] Consider AcquireAtCycle

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -250,7 +239,7 @@ void InstructionInfoView::collectData(
       const MCWriteProcResEntry *Index = STI.getWriteProcResBegin(&SCDesc);
       const MCWriteProcResEntry *Last = STI.getWriteProcResEnd(&SCDesc);
-      auto Sep = "";
+      StringRef Sep = "";
```
Reading your code again, I think you can use `ListSeparator` here: https://github.com/llvm/llvm-project/blob/617646ab8d148ab51e8b94ccc0a40af2fc6c3dcf/llvm/include/llvm/ADT/StringExtras.h#L527

mshockwave commented on llvm/docs/CommandGuide/llvm-mca.rst:
```diff
@@ -216,6 +216,13 @@ option specifies "``-``", then the output will also be sent to standard output.
 
+  `<Resources units>` syntax can be:
+  - <Resource Name>: ReleaseAtCycle is 1.
+  - <Resource Name>[<ReleaseAtCycle>]: ReleaseAtCycle greater than 1
```
nit: "is greater" ditto for other occurrences

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +60,140 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+    }
+  }
+
+  std::vector<unsigned> Paddings = {0, 7, 14, 21, 28, 35};
+  std::vector<std::string> Fields = {"#uOps",       "Latency",
+                                     "RThroughput", "MayLoad",
+                                     "MayStore",    "HasSideEffects (U)"};
+  std::vector<std::string> EndFields;
```
nit: please use SmallVector here

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +60,140 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
-    else if (IIVDEntry.Latency < 100)
-      TempStream << ' ';
+  FOS << "\n\nInstruction Info:\n";
+  for (unsigned i = 0, N = Fields.size(); i < N; i++) {
```
remove curly braces

JulienVillette committed:
[MCA] Use ListSeparator and SmallVector

topperc commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -12,15 +12,47 @@
+  // '#' comment mark is not supported by llvm-mca
+
+  CommentString = "";
+  if (Pos = S.find("\n"); Pos != std::string::npos) {
```
StringRef::npos since you're using StringRef::find

topperc commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -29,82 +61,139 @@ void InstructionInfoView::printView(raw_ostream &OS) const {
+                                       "RThroughput", "MayLoad",
+                                       "MayStore",    "HasSideEffects (U)"};
+  SmallVector<StringRef, 8> EndFields;
+  unsigned LastPadding = 35;
```
LastPadding = Paddings.back()

JulienVillette committed:
[MCA] Fix StringRef::npos and get last padding using back()

RKSimon commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-sve-instructions.s:
```diff
@@ -2467,2480 +2467,2504 @@ zip2	z31.d, z31.d, z31.d
-# CHECK-NEXT:  1      2     0.50                        zip2	z31.d, z31.d, z31.d
-# CHECK-NEXT:  1      2     0.50                        zip2	z31.h, z31.h, z31.h
-# CHECK-NEXT:  1      2     0.50                        zip2	z31.s, z31.s, z31.s
+# CHECK:      [1]    [2]    [3]    [4]    [5]    [6]    [7]    [8]                                        [9]                        Instructions:
```
(pedantic) left align columns [8] and [9]?

RKSimon commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -12,15 +12,47 @@
 namespace mca {
 
+void InstructionInfoView::getComment(const MCInst &MCI,
+                                     std::string &CommentString) const {
```
Would it be better just to write to a raw_ostream (like we do for dump calls)?

RKSimon commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,32 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
+               clEnumValN(InstructionTablesType::FULL, "full",
+                          "Print instruction tables with additional"
+                          " information: bypass latency, LLVM opcode,"
+                          " used resources")),
```
can we add test coverage for all of these options?

jvillette38 replied:
Will add ./llvm/test/tools/llvm-mca/RISCV/SiFive7/instruction-tables-tests.s

jvillette38 replied:
Add the test in RISCV/SiFive7 to get AcquireAtCycle.

RKSimon commented on llvm/tools/llvm-mca/llvm-mca.cpp:
```diff
@@ -225,10 +225,32 @@ static cl::opt<unsigned> StoreQueueSize("squeue",
+  if (InstructionTablesOption == ITType)
+    return true;
+
+  return false;
```
`return InstructionTablesOption == ITType;` ?

RKSimon reviewed (commented):
a few minors

JulienVillette committed:
[MCA] Fixing formatting

JulienVillette committed:
[MCA] Fix clang-format

mshockwave commented on llvm/test/tools/llvm-mca/RISCV/SiFive7/instruction-tables-tests.s:
```diff
@@ -0,0 +1,415 @@
+# RUN: llvm-mca -march=riscv64 -mcpu=sifive-u74 -mattr=+v -instruction-tables=full -show-encoding < %s | FileCheck %s -check-prefixes=ALL,ISFE
+# RUN: llvm-mca -march=riscv64 -mcpu=sifive-u74 -mattr=+v -instruction-tables=full -show-barriers -show-encoding < %s | FileCheck %s -check-prefixes=ALL,ISFBE
+
+  vsetvli a3, a2, e16, m1, tu, mu   // Comment
```
what is the "// Comment" at the end for?

jvillette38 replied:
Example of input:
`abs D2, D19  // ABS <V><d>, <V><n>  \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV`
1: micro ops, 2: latency, 2: bypass latency, 4.0: throughput (not reverse), Units with release at cycle if needed.
Output:
` 1  2  0.25  2  V1UnitV   ABSv1i64   abs  d2, d19 // ABS <V><d>, <V><n>  \\ ASIMD arith, basic  \\ 1 2  2  4.0 V1UnitV`

I can then use a script to check automatically values in comments compared to values returned by llvm-mca.

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -12,15 +12,45 @@
+      return;
+    }
+  }
+  return;
```
redundant return

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -12,15 +12,45 @@
+    // C++ style comment
+    if ((PosCmt = InstrStr.find("//")) != StringRef::npos) {
+      OS << InstrStr.substr(PosCmt);
+      return;
```
redundant return

mshockwave commented on llvm/tools/llvm-mca/Views/InstructionInfoView.cpp:
```diff
@@ -12,15 +12,45 @@
+  // '#' comment mark is not supported by llvm-mca
+
+  if (Pos = S.find("\n"); Pos != StringRef::npos) {
+    InstrStr = S.take_front(Pos);
```
please move the declaration here, as you're not using `InstrStr` outside the encoding if-statement

mshockwave commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V1-sve-instructions.s:
```diff
@@ -1,7 +1,7 @@
+# RUN: llvm-mca -mtriple=aarch64 -mcpu=neoverse-v1 -instruction-tables=full < %s | FileCheck %s
 
-abs	z0.b, p0/m, z0.b
+abs	z0.b, p0/m, z0.b    // Toto
```
what is "// Toto" at the end for?

jvillette38 replied:
It was just to test comment but already done in `RISCV/SiFive7/instruction-tables-tests.s`

JulienVillette committed:
[MCA] Code review remarks

mshockwave reviewed (approved):
LGTM cheers!

jvillette38 commented:
I have created new MR #132972 with the fix (just pushed not yet validated).
@ilovepi, as you want!

mshockwave commented:
> I have created new MR #132972 with the fix (just pushed not yet validated). @ilovepi, as you want! Thanks!

FYI: I just merged the fix.

jvillette38 commented:
> > I have created new MR #132972 with the fix (just pushed not yet validated). @ilovepi, as you want! Thanks!
> 
> FYI: I just merged the fix.

Thanks! It seems to be fixed now.
